docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,423 @@
+currently manually written to expose additional AKS API fields to CAPZ.
+
+Once implemented, this method would drastically reduce the amount of developer effort to get started adding
+new AKS features. There would continue to be some amount of ongoing cost though to identify and handle nuances
+for each feature though, which may include issues outside of CAPZ's control like AKS API quirks.
+
+The main drawback of this approach from a developer perspective is the up-front effort to implement and
+ongoing cost to maintain the code generation itself. The existing AKS feature set exposed by CAPZ provides a
+decent foundation that can help identify regressions by testing the pipeline against features already
+implemented and tested. ASO also already implements a full code generation pipeline to transform Azure API
+specs into Kubernetes resource definitions, so some of that could possibly be reused by CAPZ.
+
+#### Option 7: CAPZ resource defines patches to ASO resource
+
+This method describes adding a new `spec.asoManagedClusterPatches` field to the existing
+AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+
+	// ASOManagedClusterPatches defines patches to be applied to the generated ASO ManagedCluster resource.
+	ASOManagedClusterPatches []string `json:"asoManagedClusterPatches,omitempty"`
+}
+```
+
+After CAPZ calculates the ASO ManagedCluster for a given AzureManagedControlPlane during a reconciliation, it
+will apply these patches in order to the ManagedCluster before ultimately submitting that to ASO. This allows
+users to specify any AKS feature declaratively within the existing CAPZ spec. The exact format of the patches
+is TBD, but likely one or all of the formats described here:
+https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/
+
+The main drawback of this approach is the fragility of the patches themselves, which can break underneath
+users with a new version of ASO or CAPZ. There is also increased risk that a user could modify the ASO
+resource in a way that breaks CAPZ's ability to reconcile it, though CAPZ could perhaps add some extra
+validation that particularly sensitive fields do not get modified by a set of patches. Compared to other
+options, the syntax for specifying a patch is also more cumbersome than if its equivalent had its own CAPZ API
+field. Given the intent of setting patches is to enable niche use-cases for advanced users, these drawbacks
+may be acceptable.
+
+#### Option 8: Users bring-their-own ASO ManagedCluster resource
+
+CAPZ can already incorporate existing ASO resources into Cluster API Cluster configurations. BYO ASO resources
+are never modified or deleted by CAPZ directly but are still read and play in to the status of CAPZ resources.
+
+The flow for this method is roughly:
+1. User creates an ASO ManagedCluster before or at the same time as the other Cluster API and CAPZ resources.
```

Comment:
I assume this would require some subtle tweaking to the existing CAPZ AzureManagedControlPlane reconciliation, as there could be a race between the user-provided ASO ManagedCluster and CAPZ's detection that an existing ASO ManagedCluster reference already exists?

Reply:
Perhaps. I think the best practice for this would be for users to synchronize that themselves and ensure the ASO ManagedCluster is created before the AzureManagedControlPlane. Otherwise we'd still need to invent some mechanism for users to declaratively specify "don't create this ASO resource for me because I'll do it myself."

In practice though, I think it would be safe to apply all the YAML at once assuming CAPZ isn't fast enough to create a ManagedCluster within a few dozen milliseconds or whatever of the AzureManagedControlPlane hitting the API server.

Reply:
As dumb as it sounds IME the ordering of the YAML is respected by the apiserver. So CAPZ's GET should always return a non-404 response when it does its thing at the beginning of the AzureManagedControlPlane create reconciliation flow.

Reply:
Is CAPZ using a cached or non-cached client for those GETs? If it's a cached client w/ a watch it'll return NotFound until such time as the apiserver sends an update to the watch and informs CAPZ of the creation of the cluster. This can definitely happen (I've seen it in ASO before, though I forget the specific context).